File: chargesearch.cs
// Copyright 2025 William Stafford Parsons

public static int ChargeSearch(int[] haystack, int needle) {
  int high = (haystack.Length | 1) ^ 1;
  int i = 0;

  while (i < high) {
    if (haystack[i] == needle) {
      return i;
    }

    if (haystack[i + 1] == needle) {
      return i + 1;
    }

    i += 2;
  }

  if (
    i < haystack.Length &&
    haystack[i] == needle
  ) {
    return i;
  }

  return -1;
}

File: chargesearch.go
// Copyright 2025 William Stafford Parsons

func chargeSearch(haystack []int, needle int) int {
  high := (len(haystack) | 1) ^ 1
  i := 0

  for i < high {
    if haystack[i] == needle {
      return i
    }

    if haystack[i + 1] == needle {
      return i + 1
    }

    i += 2
  }

  if i < len(haystack) && haystack[i] == needle {
    return i
  }

  return -1
}
